In [1]:
import numpy as np
import pandas as pd

In [2]:
bike_df=pd.read_csv('Used_Bikes.csv')

In [3]:
bike_df.shape

(32648, 8)

In [4]:
bike_df=bike_df[bike_df['price']<500000]

In [5]:
bike_df.shape

(32495, 8)

In [7]:
bike_df.head()

bike_name     price       city  kms_driven  \
0    TVS Star City Plus Dual Tone 110cc   35000.0  Ahmedabad     17654.0   
1           Royal Enfield Classic 350cc  119900.0      Delhi     11000.0   
3                  TVS Apache RTR 180cc   65000.0  Bangalore     16329.0   
4  Yamaha FZ S V 2.0 150cc-Ltd. Edition   80000.0  Bangalore     10000.0   
5                      Yamaha FZs 150cc   53499.0      Delhi     25000.0   

         owner  age  power          brand  
0  First Owner  3.0  110.0            TVS  
1  First Owner  4.0  350.0  Royal Enfield  
3  First Owner  4.0  180.0            TVS  
4  First Owner  3.0  150.0         Yamaha  
5  First Owner  6.0  150.0         Yamaha

In [9]:
import re


In [10]:
def bike_name_cleaner(obj):
    pattern = r'[0-9]'
    mod_string = re.sub(pattern, '', obj)
    mod_string= mod_string.replace('cc','')
    return mod_string
  

In [11]:
bike_df['bike_name']=bike_df['bike_name'].apply(bike_name_cleaner)

In [12]:
bike_df.head(20)

bike_name     price       city  kms_driven  \
0         TVS Star City Plus Dual Tone    35000.0  Ahmedabad     17654.0   
1                Royal Enfield Classic   119900.0      Delhi     11000.0   
3                       TVS Apache RTR    65000.0  Bangalore     16329.0   
4         Yamaha FZ S V . -Ltd. Edition   80000.0  Bangalore     10000.0   
5                           Yamaha FZs    53499.0      Delhi     25000.0   
6            Honda CB Hornet R  ABS DLX   85000.0      Delhi      8200.0   
7        Hero Splendor Plus Self Alloy    45000.0      Delhi     12645.0   
8          Royal Enfield Thunderbird X   145000.0  Bangalore      9190.0   
9   Royal Enfield Classic Desert Storm    88000.0      Delhi     19000.0   
10                      Yamaha YZF-R .    72000.0  Bangalore     20000.0   
11                           Yamaha FZ    95000.0  Bangalore      9665.0   
12                      Bajaj Pulsar NS   78000.0  Bangalore      9900.0   
13                     Bajaj Discover M   29499.0      Delhi     20000.0   
14                     Bajaj Discover M   29900.0      Delhi     20000.0   
15                  Bajaj Pulsar NS ABS   90000.0  Bangalore     11574.0   
16                  Bajaj Pulsar RS ABS  120000.0  Bangalore     23000.0   
17                    Suzuki Gixxer SF    48000.0     Mumbai     24725.0   
18                         Benelli R CC  240000.0     Mumbai     15025.0   
19                     Bajaj Discover M   29900.0      Delhi     20000.0   
20                  Bajaj Pulsar RS ABS  120000.0  Bangalore     23000.0   

           owner  age  power          brand  
0    First Owner  3.0  110.0            TVS  
1    First Owner  4.0  350.0  Royal Enfield  
3    First Owner  4.0  180.0            TVS  
4    First Owner  3.0  150.0         Yamaha  
5    First Owner  6.0  150.0         Yamaha  
6    First Owner  3.0  160.0          Honda  
7    First Owner  3.0  100.0           Hero  
8    First Owner  3.0  350.0  Royal Enfield  
9   Second Owner  7.0  500.0  Royal Enfield  
10   First Owner  7.0  150.0         Yamaha  
11   First Owner  4.0  250.0         Yamaha  
12   First Owner  4.0  200.0          Bajaj  
13   First Owner  8.0  100.0          Bajaj  
14   First Owner  7.0  125.0          Bajaj  
15   First Owner  3.0  200.0          Bajaj  
16   First Owner  3.0  200.0          Bajaj  
17   First Owner  5.0  150.0         Suzuki  
18  Second Owner  3.0  302.0        Benelli  
19   First Owner  7.0  125.0          Bajaj  
20   First Owner  3.0  200.0          Bajaj

In [10]:
bike_df=bike_df.drop(columns=['owner'])

In [11]:
bike_df.head()

bike_name     price       city  kms_driven  age  power  \
0  TVS Star City Plus Dual Tone    35000.0  Ahmedabad     17654.0  3.0  110.0   
1         Royal Enfield Classic   119900.0      Delhi     11000.0  4.0  350.0   
3                TVS Apache RTR    65000.0  Bangalore     16329.0  4.0  180.0   
4  Yamaha FZ S V . -Ltd. Edition   80000.0  Bangalore     10000.0  3.0  150.0   
5                    Yamaha FZs    53499.0      Delhi     25000.0  6.0  150.0   

           brand  
0            TVS  
1  Royal Enfield  
3            TVS  
4         Yamaha  
5         Yamaha

### split

split the data transfrom the data( encode coloumn) apply model make pipeline and fit it dump pipeline

In [12]:
X=bike_df.drop(columns=['price'])


In [13]:
X.head()

bike_name       city  kms_driven  age  power  \
0  TVS Star City Plus Dual Tone   Ahmedabad     17654.0  3.0  110.0   
1         Royal Enfield Classic       Delhi     11000.0  4.0  350.0   
3                TVS Apache RTR   Bangalore     16329.0  4.0  180.0   
4  Yamaha FZ S V . -Ltd. Edition  Bangalore     10000.0  3.0  150.0   
5                    Yamaha FZs       Delhi     25000.0  6.0  150.0   

           brand  
0            TVS  
1  Royal Enfield  
3            TVS  
4         Yamaha  
5         Yamaha

In [14]:
y=bike_df['price']

In [15]:
y.head()

0     35000.0
1    119900.0
3     65000.0
4     80000.0
5     53499.0
Name: price, dtype: float64

In [16]:
from sklearn.preprocessing import OneHotEncoder
from sklearn.model_selection import train_test_split
from sklearn.compose import make_column_transformer
from sklearn.pipeline import make_pipeline

In [17]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=0)
X_train.head()

bike_name         city  kms_driven   age  power  \
30543  Royal Enfield Bullet Electra         Delhi     20000.0   4.0  350.0   
8936               Bajaj Pulsar  NS      Vadodara     28000.0  13.0  200.0   
20756                  Yamaha SZ-RR   Kanchipuram     52000.0  10.0  150.0   
654                 Jawa Standard CC    Bangalore      1255.0   2.0  295.0   
31859          Bajaj Avenger Street        Mumbai      8000.0   5.0  220.0   

               brand  
30543  Royal Enfield  
8936           Bajaj  
20756         Yamaha  
654             Jawa  
31859          Bajaj

In [18]:
ohe=OneHotEncoder()
ohe.fit(X[['bike_name','city','brand']])

OneHotEncoder()

#### column transformer

one hot encoding -> bike_name, city, brand ordinal encoding -> owner

In [19]:
step1= make_column_transformer((OneHotEncoder(categories=ohe.categories_,sparse=False,drop="first"),[0,1,5])
                              ,remainder='passthrough'
)

In [20]:
from sklearn.linear_model import LinearRegression

In [21]:
step2=LinearRegression()

In [22]:
pipe=make_pipeline(step1,step2)

In [23]:
pipe.fit(X_train,y_train)

Pipeline(steps=[('columntransformer',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('onehotencoder',
                                                  OneHotEncoder(categories=[array(['BMW G  GS', 'BMW G  R', 'Bajaj  Pulsar ', 'Bajaj Avenger ',
       'Bajaj Avenger Cruise ', 'Bajaj Avenger Cruise  ABS',
       'Bajaj Avenger Street ', 'Bajaj Avenger Street  ABS',
       'Bajaj Boxer AT', 'Bajaj Boxer BM', 'Bajaj Boxer CT', 'Bajaj CT  ',
       'Bajaj CT  Alloy', '...
       'Virudhunagar', 'Visakhapatnam', 'Vizianagaram', 'Warangal',
       'Wardha', 'Yamuna Nagar', 'Yemmiganur', 'Zirakpur'], dtype=object),
                                                                            array(['BMW', 'Bajaj', 'Benelli', 'Ducati', 'Harley-Davidson', 'Hero',
       'Honda', 'Hyosung', 'Ideal', 'Jawa', 'KTM', 'Kawasaki', 'LML',
       'Mahindra', 'Rajdoot', 'Royal Enfield', 'Suzuki', 'TVS', 'Yamaha',
       'Yezdi'], dtype=object

In [24]:
from sklearn.metrics import r2_score,mean_absolute_error

In [25]:
y_pred=pipe.predict(X_test)
print('r2score',r2_score(y_test,y_pred))
print('maescore',mean_absolute_error(y_test,y_pred))


r2score 0.9787799778089978
maescore 4071.083235045315


### result

the model has 98% accuracy and has 4071 in price difference to orignal price

In [30]:
pipe.predict(pd.DataFrame([['Yamaha FZ','Faridabad',100000,8,150,'Yamaha']],columns=['bike_name','city','kms_driven','age','power','brand']))

ValueError: Found unknown categories ['Yamaha FZ'] in column 0 during transform

In [28]:
X_train.head(20)

bike_name         city  kms_driven   age  power  \
30543  Royal Enfield Bullet Electra         Delhi     20000.0   4.0  350.0   
8936               Bajaj Pulsar  NS      Vadodara     28000.0  13.0  200.0   
20756                  Yamaha SZ-RR   Kanchipuram     52000.0  10.0  150.0   
654                 Jawa Standard CC    Bangalore      1255.0   2.0  295.0   
31859          Bajaj Avenger Street        Mumbai      8000.0   5.0  220.0   
24782                     Yamaha FZ       Chennai     18742.0   6.0  150.0   
13901                   Suzuki Zeus     Ahmedabad     11885.0  12.0  125.0   
27776                  Bajaj Pulsar     Ghaziabad     64955.0  15.0  150.0   
7991                 Bajaj Pulsar AS        Delhi     55250.0   6.0  200.0   
14804    Harley-Davidson Street  ABS      Jodhpur      3430.0   4.0  750.0   
18849              Hero Passion Pro         Delhi     22000.0   4.0  100.0   
28200           Hero Super Splendor        Jaipur     84186.0  16.0  125.0   
7065      Royal Enfield Thunderbird         Surat     27500.0   6.0  350.0   
4430                 Honda CB Shine        Jaipur     22317.0   8.0  125.0   
23296                TVS Apache RTR        Karnal      6639.0   9.0  180.0   
11276               Hero CBZ Xtreme         Noida     34968.0  14.0  150.0   
6837         Royal Enfield Standard      Ludhiana     25000.0  18.0  350.0   
5782                     TVS Apache   Kanchipuram      6807.0  14.0  150.0   
12567                Yamaha YZF-R .      Rupnagar     27000.0   9.0  150.0   
22055                 Yamaha Fazer         Kadapa     14500.0   4.0  250.0   

                 brand  
30543    Royal Enfield  
8936             Bajaj  
20756           Yamaha  
654               Jawa  
31859            Bajaj  
24782           Yamaha  
13901           Suzuki  
27776            Bajaj  
7991             Bajaj  
14804  Harley-Davidson  
18849             Hero  
28200             Hero  
7065     Royal Enfield  
4430             Honda  
23296              TVS  
11276             Hero  
6837     Royal Enfield  
5782               TVS  
12567           Yamaha  
22055           Yamaha

In [27]:
import pickle

In [28]:
pickle.dump(pipe,open('bike_pipe.pkl','wb'))
pickle.dump(bike_df,open('bike_new1_df.pkl','wb'))

In [28]:
X_train.head()

bike_name         city  kms_driven   age  power  \
30543  Royal Enfield Bullet Electra         Delhi     20000.0   4.0  350.0   
8936               Bajaj Pulsar  NS      Vadodara     28000.0  13.0  200.0   
20756                  Yamaha SZ-RR   Kanchipuram     52000.0  10.0  150.0   
654                 Jawa Standard CC    Bangalore      1255.0   2.0  295.0   
31859          Bajaj Avenger Street        Mumbai      8000.0   5.0  220.0   

               brand  
30543  Royal Enfield  
8936           Bajaj  
20756         Yamaha  
654             Jawa  
31859          Bajaj